### PR TITLE
Remove connectToNodeLight and replace it with a connection profile

### PR DESCRIPTION
**Pull request by s1monw**
The Transport#connectToNodeLight concepts is confusing and not very flexible.
neither really testable on a unittest level. This commit cleans up the code used
to connect to nodes and simplifies transport implementations to share more code.
This also allows to connect to nodes with custom profiles if needed, for instance
future improvements can be added to connect to/from nodes that are non-data nodes without
dedicated bulks and recovery connections.

### CODE REVIEW

**Comment by s1monw** [comment]
test this please

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java [hunk 1]
```diff
@@ -331,65 +332,29 @@ public long serverOpen() {
...
-                }
-                if (nodeChannels.recovery.length == 0) {
-                    if (nodeChannels.bulk.length > 0) {
-                        nodeChannels.recovery = nodeChannels.bulk;
-                    } else {
-                        nodeChannels.recovery = nodeChannels.reg;
+                for (int i = 0; i < channels.length; i++) {
+                    assert iterator.hasNext();
+                    ChannelFuture future = iterator.next();
+                    future.awaitUninterruptibly((long) (connectTimeout.millis() * 1.5));
```
Comment (s1monw) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java]: good idea, very good idea. I will create a followup for this WDYT @tlrx

Comment (tlrx) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java]: +1

**Comment by s1monw** [comment]
test this please

File: modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java (hunk 1 above)
Comment (s1monw) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java]: @tlrx I looked into this and one problem is that the connect timeout is per Bootstrap on the netty layer. This essentially means we need a dedicated Bootstrap for each connection timeout value. I don't think we should do that, maybe there is a better solution to this? I wonder if we should look into adding this option to netty?

Comment (tlrx) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java]: great!

Comment (s1monw) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java]: @tlrx is it [that](https://github.com/s1monw/elasticsearch/commit/485bba138ed18581c40cd01853939df54a59ee97) what you had in mind

Comment (tlrx) [inline modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4Transport.java]: @s1monw yes

**Comment by s1monw** [comment]
as a followup I'd like to do  stuff like this:
```DIFF
diff --git a/core/src/main/java/org/elasticsearch/transport/TcpTransport.java b/core/src/main/java/org/elasticsearch/transport/TcpTransport.java
index 66dd217..e0c1d3a 100644
--- a/core/src/main/java/org/elasticsearch/transport/TcpTransport.java
+++ b/core/src/main/java/org/elasticsearch/transport/TcpTransport.java
@@ -81,6 +81,7 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.EnumMap;
+import java.util.EnumSet;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Iterator;
@@ -205,9 +206,20 @@ public abstract class TcpTransport<Channel> extends AbstractLifecycleComponent i
         ConnectionProfile.Builder builder = new ConnectionProfile.Builder();
         builder.addConnections(connectionsPerNodeBulk, TransportRequestOptions.Type.BULK);
         builder.addConnections(connectionsPerNodePing, TransportRequestOptions.Type.PING);
-        builder.addConnections(connectionsPerNodeRecovery, TransportRequestOptions.Type.RECOVERY);
-        builder.addConnections(connectionsPerNodeReg, TransportRequestOptions.Type.REG);
-        builder.addConnections(connectionsPerNodeState, TransportRequestOptions.Type.STATE);
+        Set<TransportRequestOptions.Type> regSet = EnumSet.of(TransportRequestOptions.Type.REG);
+        if (DiscoveryNode.isMasterNode(settings)) {
+            builder.addConnections(connectionsPerNodeState, TransportRequestOptions.Type.STATE);
+        } else {
+            // if we are not master eligible we don't need a dedicated channel to publish the state
+            regSet.add(TransportRequestOptions.Type.STATE);
+        }
+        if (DiscoveryNode.isDataNode(settings)) {
+            builder.addConnections(connectionsPerNodeRecovery, TransportRequestOptions.Type.RECOVERY);
+        } else {
+            // if we are not a data-node we don't need any dedicated channels for recovery
+            regSet.add(TransportRequestOptions.Type.RECOVERY);
+        }
+        builder.addConnections(connectionsPerNodeReg, regSet.toArray(new TransportRequestOptions.Type[regSet.size()]));
         defaultConnectionProfile = builder.build();
     }

```

File: core/src/main/java/org/elasticsearch/transport/ConnectionProfile.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+         * @param numConnections the number of connections to use in the pool for the given connection types
+         * @param types a set of types that should share the given number of connections
+         */
+        public void addConnections(int numConnections, TransportRequestOptions.Type... types) {
+            if (types == null || types.length == 0) {
+                throw new IllegalArgumentException("types must not be null");
+            }
+            for (TransportRequestOptions.Type type : types) {
+                if (addedTypes.contains(type)) {
+                    throw new IllegalArgumentException("type [" + type + "] is already registered");
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/transport/ConnectionProfile.java]: I think that this should be an `IllegalStateException`.

File: core/src/main/java/org/elasticsearch/transport/ConnectionProfile.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+                }
+            }
+            addedTypes.addAll(Arrays.asList(types));
+            handles.add(new ConnectionTypeHandle(offset, numConnections, types));
+            offset += numConnections;
+        }
+
+        public ConnectionProfile build() {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/transport/ConnectionProfile.java]: Can you add a Javadoc for this documenting that all types have to have been added?

**Review by jasontedor — CHANGES_REQUESTED** [review]
I like it @s1monw. I left a few comments.

File: core/src/main/java/org/elasticsearch/transport/ConnectionProfile.java (hunk 2 above)
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/transport/ConnectionProfile.java]: my reasoning is that the argument would introduce this state but I won't let it hence the ISE

**Comment by s1monw** [comment]
@jasontedor I pushed a new commit and replied to one comment

File: core/src/main/java/org/elasticsearch/transport/ConnectionProfile.java (hunk 2 above)
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/transport/ConnectionProfile.java]: Okay.

File: core/src/main/java/org/elasticsearch/transport/ConnectionProfile.java [hunk 4]
```diff
@@ -79,6 +79,10 @@ public void addConnections(int numConnections, TransportRequestOptions.Type... t
             offset += numConnections;
         }
 
+        /**
+         * Creates a new {@link ConnectionProfile} based on the added connections.
+         * @throws IllegalArgumentException if any of the {@link org.elasticsearch.transport.TransportRequestOptions.Type} enum is missing
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/transport/ConnectionProfile.java]: This Javadoc is incorrect; the method throws an `IllegalStateException` (which in this case I think is correct). :smile:

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/transport/ConnectionProfile.java]: This Javadoc is incorrect; the method throws an `IllegalStateException` (which in this case I think is correct). :smile:

**Review by jasontedor — COMMENTED** [review]
I left one more comment @s1monw.

**Comment by s1monw** [comment]
@jasontedor fixed the javadocs issue

**Review by jasontedor — APPROVED** [review]
LGTM.

**Pull request merged**